File: packages/spreadsheet-engine/src/styles.ts
import type { SpreadsheetCellStyle, SpreadsheetColor } from "./model.ts";
import { attributes, escapeXml } from "./xml.ts";

type StyleReference = Readonly<{
  borderId: number;
  fillId: number;
  fontId: number;
  numFmtId: number;
}>;

const BUILTIN_FORMATS: Readonly<Record<number, string>> = {
  0: "General",
  1: "0",
  2: "0.00",
  9: "0%",
  10: "0.00%",
  14: "m/d/yy",
  49: "@",
};

function color(value: string | undefined): SpreadsheetColor | undefined {
  if (!value) return undefined;
  const normalized = value.replace(/^#/u, "").toUpperCase();
  if (/^[\dA-F]{8}$/u.test(normalized)) return `#${normalized.slice(2)}`;
  if (/^[\dA-F]{6}$/u.test(normalized)) return `#${normalized}`;
  return undefined;
}

function childElements(source: string, tag: string): readonly string[] {
  const block = new RegExp(`<${tag}\\b[^>]*>([\\s\\S]*?)<\\/${tag}>`, "iu").exec(source)?.[1];
  if (!block) return [];
  const singular = tag === "cellXfs" ? "xf" : tag.slice(0, -1);
  const expression =
    singular === "xf"
      ? /<xf\b(?:[^>]*\/>|[^>]*>[\s\S]*?<\/xf>)/giu
      : new RegExp(`<${singular}\\b[^>]*>[\\s\\S]*?<\\/${singular}>`, "giu");
  return [...block.matchAll(expression)].map((match) => match[0]);
}

function tagAttributes(source: string, tag: string): Readonly<Record<string, string>> {
  return attributes(new RegExp(`<${tag}\\b([^>]*)`, "iu").exec(source)?.[1] ?? "");
}

function parseFont(source: string): SpreadsheetCellStyle {
  const colorAttributes = tagAttributes(source, "color");
  const fontColor = color(colorAttributes.rgb);
  const fontFamily = tagAttributes(source, "name").val;
  const fontSize = Number(tagAttributes(source, "sz").val);
  return {
    ...(source.includes("<b") ? { bold: true } : {}),
    ...(fontColor ? { color: fontColor } : {}),
    ...(fontFamily ? { fontFamily } : {}),
    ...(fontSize > 0 ? { fontSize } : {}),
    ...(source.includes("<i") ? { italic: true } : {}),
    ...(source.includes("<u") ? { underline: true } : {}),
  };
}

function parseFill(source: string): SpreadsheetCellStyle {
  const foreground = color(tagAttributes(source, "fgColor").rgb);
  return foreground ? { background: foreground } : {};
}

function parseBorder(source: string): SpreadsheetCellStyle {
  const sideColor = (side: string): SpreadsheetColor | undefined => {
    const block = new RegExp(`<${side}\\b[^>]*>([\\s\\S]*?)<\\/${side}>`, "iu").exec(source)?.[0];
    return block ? color(tagAttributes(block, "color").rgb) : undefined;
  };
  const bottom = sideColor("bottom");
  const left = sideColor("left");
  const right = sideColor("right");
  const top = sideColor("top");
  return {
    ...(bottom ? { borderBottom: bottom } : {}),
    ...(left ? { borderLeft: left } : {}),
    ...(right ? { borderRight: right } : {}),
    ...(top ? { borderTop: top } : {}),
  };
}

function fontXml(style: SpreadsheetCellStyle): string {
  return `<font>${style.bold ? "<b/>" : ""}${style.italic ? "<i/>" : ""}${style.underline ? "<u/>" : ""}${style.fontSize ? `<sz val="${style.fontSize}"/>` : ""}${style.color ? `<color rgb="FF${style.color.slice(1)}"/>` : ""}${style.fontFamily ? `<name val="${escapeXml(style.fontFamily)}"/>` : ""}</font>`;
}

function fillXml(style: SpreadsheetCellStyle): string {
  return style.background
    ? `<fill><patternFill patternType="solid"><fgColor rgb="FF${style.background.slice(1)}"/><bgColor indexed="64"/></patternFill></fill>`
    : `<fill><patternFill patternType="none"/></fill>`;
}

function borderXml(style: SpreadsheetCellStyle): string {
  const side = (name: string, value: SpreadsheetColor | undefined) =>
    value ? `<${name} style="thin"><color rgb="FF${value.slice(1)}"/></${name}>` : `<${name}/>`;
  return `<border>${side("left", style.borderLeft)}${side("right", style.borderRight)}${side("top", style.borderTop)}${side("bottom", style.borderBottom)}<diagonal/></border>`;
}

function differentialXml(style: SpreadsheetCellStyle): string {
  return `<dxf>${Object.keys(style).some((key) => ["bold", "color", "fontFamily", "fontSize", "italic", "underline"].includes(key)) ? fontXml(style) : ""}${style.background ? fillXml(style) : ""}${Object.keys(style).some((key) => key.startsWith("border")) ? borderXml(style) : ""}${style.numberFormat ? `<numFmt numFmtId="0" formatCode="${escapeXml(style.numberFormat)}"/>` : ""}</dxf>`;
}

function appendItems(
  source: string,
  collection: string,
  item: string,
  items: readonly string[],
): string {
  if (items.length === 0) return source;
  const expression = new RegExp(`<${collection}\\b([^>]*)>([\\s\\S]*?)<\\/${collection}>`, "iu");
  return source.replace(expression, (_whole, attributeSource: string, content: string) => {
    const attrs = attributes(attributeSource);
    const count = Number(
      attrs.count || [...content.matchAll(new RegExp(`<${item}\\b`, "giu"))].length,
    );
    const nextAttributes = /\bcount\s*=/iu.test(attributeSource)
      ? attributeSource.replace(/\bcount\s*=\s*(["'])\d+\1/iu, `count="${count + items.length}"`)
      : `${attributeSource} count="${count + items.length}"`;
    return `<${collection}${nextAttributes}>${content}${items.join("")}</${collection}>`;
  });
}

export class SpreadsheetStyleStore {
  readonly #borders: readonly SpreadsheetCellStyle[];
  readonly #fills: readonly SpreadsheetCellStyle[];
  readonly #fonts: readonly SpreadsheetCellStyle[];
  readonly #formats: Map<number, string>;
  readonly #references: readonly StyleReference[];
  readonly #resolved: SpreadsheetCellStyle[];
  readonly #source: string;
  readonly #addedBorders: string[] = [];
  readonly #addedFills: string[] = [];
  readonly #addedFonts: string[] = [];
  readonly #addedFormats: string[] = [];
  readonly #addedXfs: string[] = [];
  readonly #addedDxfs: string[] = [];
  readonly #differentials: SpreadsheetCellStyle[];
  readonly #registered = new Map<string, number>();
  readonly #registeredDifferentials = new Map<string, number>();

  constructor(source: string) {
    this.#source = source;
    this.#fonts = childElements(source, "fonts").map(parseFont);
    this.#fills = childElements(source, "fills").map(parseFill);
    this.#borders = childElements(source, "borders").map(parseBorder);
    this.#formats = new Map(
      Object.entries(BUILTIN_FORMATS).map(([id, format]) => [Number(id), format]),
    );
    for (const numFmt of source.matchAll(/<numFmt\b([^>]*)\/?\s*>/giu)) {
      const attrs = attributes(numFmt[1]);
      if (attrs.numFmtId && attrs.formatCode)
        this.#formats.set(Number(attrs.numFmtId), attrs.formatCode);
    }
    const xfs = childElements(source, "cellXfs");
    this.#references = xfs.map((xf) => {
      const attrs = attributes(/^<xf\b([^>]*)/iu.exec(xf)?.[1] ?? "");
      return {
        borderId: Number(attrs.borderId ?? 0),
        fillId: Number(attrs.fillId ?? 0),
        fontId: Number(attrs.fontId ?? 0),
        numFmtId: Number(attrs.numFmtId ?? 0),
      };
    });
    this.#resolved = xfs.map((xf, index): SpreadsheetCellStyle => {
      const reference = this.#references[index];
      const alignment = tagAttributes(xf, "alignment");
      const numberFormat = this.#formats.get(reference.numFmtId);
      return {
        ...this.#fonts[reference.fontId],
        ...this.#fills[reference.fillId],
        ...this.#borders[reference.borderId],
        ...(numberFormat ? { numberFormat } : {}),
        ...(alignment.horizontal === "center" || alignment.horizontal === "right"
          ? { horizontal: alignment.horizontal }
          : alignment.horizontal === "left"
            ? { horizontal: "left" as const }
            : {}),
        ...(alignment.vertical === "center"
          ? { vertical: "middle" as const }
          : alignment.vertical === "top" || alignment.vertical === "bottom"
            ? { vertical: alignment.vertical }
            : {}),
        ...(alignment.wrapText === "1" ? { wrapText: true } : {}),
      };
    });
    if (this.#resolved.length === 0) this.#resolved.push({ numberFormat: "General" });
    this.#differentials = childElements(source, "dxfs").map((dxf) => ({
      ...parseFont(dxf),
      ...parseFill(dxf),
      ...parseBorder(dxf),
      ...(tagAttributes(dxf, "numFmt").formatCode
        ? { numberFormat: tagAttributes(dxf, "numFmt").formatCode }
        : {}),
    }));
    this.#differentials.forEach((style, index) => {
      this.#registeredDifferentials.set(this.#canonical(style), index);
    });
  }

  resolve(styleId: number): SpreadsheetCellStyle {
    return this.#resolved[styleId] ?? this.#resolved[0] ?? {};
  }

  register(style: SpreadsheetCellStyle): number {
    const canonical = JSON.stringify(
      Object.fromEntries(Object.entries(style).sort(([a], [b]) => a.localeCompare(b))),
    );
    const existing = this.#registered.get(canonical);
    if (existing !== undefined) return existing;
    const fontId = this.#fonts.length + this.#addedFonts.length;
    const fillId = this.#fills.length + this.#addedFills.length;
    const borderId = this.#borders.length + this.#addedBorders.length;
    let numFmtId =
      [...this.#formats.keys()].find((id) => this.#formats.get(id) === style.numberFormat) ?? 0;
    if (style.numberFormat && numFmtId === 0 && style.numberFormat !== "General") {
      numFmtId =
        Math.max(
          164,
          ...this.#formats.keys(),
          ...this.#addedFormats.map((item) => Number(attributes(item).numFmtId)),
        ) + 1;
      this.#formats.set(numFmtId, style.numberFormat);
      this.#addedFormats.push(
        `<numFmt numFmtId="${numFmtId}" formatCode="${escapeXml(style.numberFormat)}"/>`,
      );
    }
    this.#addedFonts.push(fontXml(style));
    this.#addedFills.push(fillXml(style));
    this.#addedBorders.push(borderXml(style));
    const alignment = [
      style.horizontal ? `horizontal="${style.horizontal}"` : "",
      style.vertical ? `vertical="${style.vertical === "middle" ? "center" : style.vertical}"` : "",
      style.wrapText ? 'wrapText="1"' : "",
    ]
      .filter(Boolean)
      .join(" ");
    this.#addedXfs.push(
      `<xf numFmtId="${numFmtId}" fontId="${fontId}" fillId="${fillId}" borderId="${borderId}" xfId="0" applyFont="1" applyFill="1" applyBorder="1" applyNumberFormat="1"${alignment ? ` applyAlignment="1"><alignment ${alignment}/></xf>` : "/>"}`,
    );
    const id = this.#resolved.length;
    this.#resolved.push(style);
    this.#registered.set(canonical, id);
    return id;
  }

  resolveDifferential(styleId: number): SpreadsheetCellStyle {
    return this.#differentials[styleId] ?? {};
  }

  registerDifferential(style: SpreadsheetCellStyle): number {
    const existing = this.#registeredDifferentials.get(this.#canonical(style));
    if (existing !== undefined) return existing;
    const id = this.#differentials.length;
    this.#differentials.push(style);
    this.#addedDxfs.push(differentialXml(style));
    this.#registeredDifferentials.set(this.#canonical(style), id);
    return id;
  }

  #canonical(style: SpreadsheetCellStyle): string {
    return JSON.stringify(
      Object.fromEntries(
        Object.entries(style).sort(([left], [right]) => left.localeCompare(right)),
      ),
    );
  }

  serialize(): string {
    let result = this.#source;
    result = appendItems(result, "fonts", "font", this.#addedFonts);
    result = appendItems(result, "fills", "fill", this.#addedFills);
    result = appendItems(result, "borders", "border", this.#addedBorders);
    result = appendItems(result, "cellXfs", "xf", this.#addedXfs);
    if (this.#addedDxfs.length > 0) {
      if (/<dxfs\b/iu.test(result)) result = appendItems(result, "dxfs", "dxf", this.#addedDxfs);
      else
        result = result.replace(
          /<\/styleSheet>\s*$/iu,
          `<dxfs count="${this.#addedDxfs.length}">${this.#addedDxfs.join("")}</dxfs></styleSheet>`,
        );
    }
    if (this.#addedFormats.length > 0) {
      if (/<numFmts\b/iu.test(result))
        result = appendItems(result, "numFmts", "numFmt", this.#addedFormats);
      else
        result = result.replace(
          /<fonts\b/iu,
          `<numFmts count="${this.#addedFormats.length}">${this.#addedFormats.join("")}</numFmts><fonts`,
        );
    }
    return result;
  }
}

File: spreadsheet-view/src/styles.rs
//! OOXML style resolution mirroring `styles.ts`: fonts, fills, borders,
//! number formats, cellXfs, and differential formats for conditional rules.

use std::collections::BTreeMap;

use crate::model::CellStyle;
use crate::xmlutil::{attributes, find_close_tag, find_open_tag};

const BUILTIN_FORMATS: &[(u32, &str)] =
    &[(0, "General"), (1, "0"), (2, "0.00"), (9, "0%"), (10, "0.00%"), (14, "m/d/yy"), (49, "@")];

/// Normalize an ARGB/RGB color to `#rrggbb`, dropping alpha exactly like the
/// engine.
pub fn normalize_color(value: Option<&str>) -> Option<String> {
    let value = value?;
    let normalized = value.trim_start_matches('#').to_uppercase();
    if normalized.len() == 8 && normalized.chars().all(|c| c.is_ascii_hexdigit()) {
        return Some(format!("#{}", &normalized[2..]));
    }
    if normalized.len() == 6 && normalized.chars().all(|c| c.is_ascii_hexdigit()) {
        return Some(format!("#{normalized}"));
    }
    None
}

fn tag_attributes(source: &str, tag: &str) -> BTreeMap<String, String> {
    match find_open_tag(source, tag, 0) {
        Some((_, attrs, _, _)) => attributes(&attrs),
        None => BTreeMap::new(),
    }
}

fn parse_font(source: &str) -> CellStyle {
    let color_attributes = tag_attributes(source, "color");
    let font_color = normalize_color(color_attributes.get("rgb").map(String::as_str));
    let font_family = tag_attributes(source, "name").get("val").cloned();
    let font_size = tag_attributes(source, "sz")
        .get("val")
        .and_then(|value| crate::jsnumber::parse(value))
        .filter(|value| *value > 0.0);
    CellStyle {
        bold: source.contains("<b").then_some(true),
        color: font_color,
        font_family: font_family.filter(|family| !family.is_empty()),
        font_size,
        italic: source.contains("<i").then_some(true),
        underline: source.contains("<u").then_some(true),
        ..Default::default()
    }
}

fn parse_fill(source: &str) -> CellStyle {
    let foreground =
        normalize_color(tag_attributes(source, "fgColor").get("rgb").map(String::as_str));
    CellStyle { background: foreground, ..Default::default() }
}

fn border_side_color(source: &str, side: &str) -> Option<String> {
    let (_, _, inner_start, self_closing) = find_open_tag(source, side, 0)?;
    if self_closing {
        return None;
    }
    let end = find_close_tag(source, side, inner_start)?;
    let block = &source[..end];
    normalize_color(tag_attributes(block, "color").get("rgb").map(String::as_str))
}

fn parse_border(source: &str) -> CellStyle {
    CellStyle {
        border_bottom: border_side_color(source, "bottom"),
        border_left: border_side_color(source, "left"),
        border_right: border_side_color(source, "right"),
        border_top: border_side_color(source, "top"),
        ..Default::default()
    }
}

fn child_elements(source: &str, collection: &str) -> Vec<String> {
    // The block between <collection ...> and </collection>.
    let (block_start, _, mut cursor, self_closing) = match find_open_tag(source, collection, 0) {
        Some(found) => found,
        None => return Vec::new(),
    };
    if self_closing {
        return Vec::new();
    }
    let block_end = match find_close_tag(source, collection, cursor) {
        Some(end) => end - collection.len() - 3,
        None => return Vec::new(),
    };
    let _ = block_start;
    let singular = if collection == "cellXfs" { "xf" } else { &collection[..collection.len() - 1] };
    let mut items = Vec::new();
    while let Some((start, _, inner_start, self_closing)) = find_open_tag(source, singular, cursor)
    {
        if start >= block_end {
            break;
        }
        let item_end = if self_closing {
            // Find the '>' that closes this singleton: re-scan from start.
            close_of_singleton(source, start)
        } else {
            find_close_tag(source, singular, inner_start).unwrap_or(inner_start)
        };
        items.push(source[start..item_end].to_string());
        cursor = item_end.max(inner_start);
    }
    items
}

fn close_of_singleton(source: &str, open_start: usize) -> usize {
    let bytes = source.as_bytes();
    let mut cursor = open_start;
    while cursor < bytes.len() {
        match bytes[cursor] {
            b'"' | b'\'' => {
                let quote = bytes[cursor];
                cursor += 1;
                while cursor < bytes.len() && bytes[cursor] != quote {
                    cursor += 1;
                }
            }
            b'>' => return cursor + 1,
            _ => cursor += 1,
        }
    }
    cursor
}

#[derive(Clone, Copy)]
struct StyleReference {
    border_id: u32,
    fill_id: u32,
    font_id: u32,
    num_fmt_id: u32,
}

fn reference_number(attrs: &BTreeMap<String, String>, key: &str) -> u32 {
    attrs
        .get(key)
        .and_then(|value| crate::jsnumber::parse(value))
        .filter(|value| value.is_finite() && *value >= 0.0 && value.fract() == 0.0)
        .map(|value| value as u32)
        .unwrap_or(0)
}

pub struct StyleStore {
    resolved: Vec<CellStyle>,
    differentials: Vec<CellStyle>,
}

impl StyleStore {
    pub fn parse(source: &str) -> Self {
        let fonts: Vec<CellStyle> =
            child_elements(source, "fonts").iter().map(|s| parse_font(s)).collect();
        let fills: Vec<CellStyle> =
            child_elements(source, "fills").iter().map(|s| parse_fill(s)).collect();
        let borders: Vec<CellStyle> =
            child_elements(source, "borders").iter().map(|s| parse_border(s)).collect();
        let mut formats: BTreeMap<u32, String> =
            BUILTIN_FORMATS.iter().map(|(id, format)| (*id, format.to_string())).collect();
        let mut cursor = 0;
        while let Some((_, attrs_source, _, _)) = find_open_tag(source, "numFmt", cursor) {
            let attrs = attributes(&attrs_source);
            if let (Some(id), Some(code)) = (
                attrs
                    .get("numFmtId")
                    .and_then(|value| crate::jsnumber::parse(value))
                    .filter(|v| v.fract() == 0.0 && *v >= 0.0),
                attrs.get("formatCode"),
            ) {
                formats.insert(id as u32, code.clone());
            }
            cursor += 1;
        }
        let xfs = child_elements(source, "cellXfs");
        let references: Vec<StyleReference> = xfs
            .iter()
            .map(|xf| {
                let attrs = match find_open_tag(xf, "xf", 0) {
                    Some((_, attrs_source, _, _)) => attributes(&attrs_source),
                    None => BTreeMap::new(),
                };
                StyleReference {
                    border_id: reference_number(&attrs, "borderId"),
                    fill_id: reference_number(&attrs, "fillId"),
                    font_id: reference_number(&attrs, "fontId"),
                    num_fmt_id: reference_number(&attrs, "numFmtId"),
                }
            })
            .collect();
        let mut resolved: Vec<CellStyle> = xfs
            .iter()
            .zip(references.iter())
            .map(|(xf, reference)| {
                let alignment = tag_attributes(xf, "alignment");
                let number_format = formats.get(&reference.num_fmt_id);
                let horizontal = alignment.get("horizontal").map(String::as_str);
                let vertical = alignment.get("vertical").map(String::as_str);
                CellStyle {
                    number_format: number_format.cloned(),
                    horizontal: horizontal
                        .filter(|value| matches!(*value, "center" | "right" | "left"))
                        .map(str::to_string),
                    vertical: vertical
                        .map(|value| match value {
                            "center" => "middle",
                            "top" | "bottom" => value,
                            _ => "",
                        })
                        .filter(|value| !value.is_empty())
                        .map(str::to_string),
                    wrap_text: (alignment.get("wrapText").map(String::as_str) == Some("1"))
                        .then_some(true),
                    ..merge_styles(
                        merge_styles(
                            fonts.get(reference.font_id as usize).cloned().unwrap_or_default(),
                            fills.get(reference.fill_id as usize).cloned().unwrap_or_default(),
                        ),
                        borders.get(reference.border_id as usize).cloned().unwrap_or_default(),
                    )
                }
            })
            .collect();
        if resolved.is_empty() {
            resolved.push(CellStyle {
                number_format: Some("General".to_string()),
                ..Default::default()
            });
        }
        let differentials: Vec<CellStyle> = child_elements(source, "dxfs")
            .iter()
            .map(|dxf| {
                let mut style = merge_styles(parse_font(dxf), parse_fill(dxf));
                style = merge_styles(style, parse_border(dxf));
                if let Some(format_code) = tag_attributes(dxf, "numFmt").get("formatCode") {
                    style.number_format = Some(format_code.clone());
                }
                style
            })
            .collect();
        drop((fonts, fills, borders, formats));
        Self { resolved, differentials }
    }

    /// The fully-resolved style for a cellXfs index.
    pub fn resolve(&self, style_id: u32) -> CellStyle {
        self.resolved
            .get(style_id as usize)
            .or_else(|| self.resolved.first())
            .cloned()
            .unwrap_or_default()
    }

    /// The differential (dxf) style used by conditional formatting rules.
    pub fn resolve_differential(&self, style_id: u32) -> CellStyle {
        self.differentials.get(style_id as usize).cloned().unwrap_or_default()
    }
}

fn merge_styles(mut base: CellStyle, update: CellStyle) -> CellStyle {
    macro_rules! overlay {
        ($($field:ident),* $(,)?) => {
            $(if update.$field.is_some() {
                base.$field = update.$field;
            })*
        };
    }
    overlay!(
        background,
        bold,
        border_bottom,
        border_left,
        border_right,
        border_top,
        color,
        font_family,
        font_size,
        horizontal,
        italic,
        number_format,
        underline,
        vertical,
        wrap_text,
    );
    base
}

/// Merge two styles with later values winning, mirroring the spread order the
/// engine used when composing font/fill/border parts.
pub fn merged_style(base: CellStyle, update: CellStyle) -> CellStyle {
    merge_styles(base, update)
}
